### PR TITLE
EWB-2453: Rework token fetcher

### PR DESCRIPTION
**Pull request by Lax125**
# Description

- Unifies verify certificate boolean and ca filename string into a single parameter that is passed into `requests.get`/`requests.post`.
- Removes default paths for auth config. The entire address is passed into `create_token_fetcher` now.
- Refactors classes to be in line with the package structure in the JVM version

# Associated tasks:

Tasks blocking this one:
- None

Tasks this is blocking:
- https://github.com/zepben/evolve-sdk-python/pull/115

# Checklist:

- [x] I have performed a self review of my own code.
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have updated the changelog.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE.
- [x] I have considered if this is a breaking change and have communicated it with other team members if so.

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-2453 Rework ZepbenTokenFetcher in Python SDK](https://app.clickup.com/t/6929263/EWB-2453)

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-2451 Port improvements from JVM connect functions to Python SDK](https://app.clickup.com/t/6929263/EWB-2451)

File: src/zepben/auth/client/zepben_token_fetcher.py [hunk 1]
```diff
@@ -5,43 +5,28 @@
...
+from zepben.auth.common.status_code import StatusCode
 
 
-class AuthException(Exception):
-    pass
-
-
-class AuthMethod(Enum):
+@dataclass
+class ZepbenTokenFetcher(object):
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/auth/client/zepben_token_fetcher.py:23]: @sonatype-lift ignore

File: test/client/test_token_fetcher.py [hunk 2]
```diff
@@ -8,11 +12,11 @@
...
 from unittest import mock
-from unittest.mock import ANY, call
+from unittest.mock import ANY
 
 import pytest
 
-from zepben.auth.client.token_fetcher import ZepbenTokenFetcher, AuthException, create_token_fetcher, AuthMethod
+from zepben.auth import ZepbenTokenFetcher, AuthException, create_token_fetcher, AuthMethod
 
 TOKEN = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImZha2VraWQifQ.eyJpc3MiOiJodHRwczovL2lzc3Vlci8iLCJzdWIiOiJmYWtlIiwiYXVkIjoiaHR0cHM6Ly9mYWtlLWF1ZC8iLCJpYXQiOjE1OTE4MzQxNzksImV4cCI6OTU5MTkyMDU3OSwiYXpwIjoid2U5ZDNSME5jTUNWckpDZ2ROSWVmWWx6aHo2VE9SaGciLCJzY29wZSI6IndyaXRlOm5ldHdvcmsgcmVhZDpuZXR3b3JrIHdyaXRlOm1ldHJpY3MgcmVhZDpld2IiLCJndHkiOiJjbGllbnQtY3JlZGVudGlhbHMiLCJwZXJtaXNzaW9ucyI6WyJ3cml0ZTpuZXR3b3JrIiwicmVhZDpuZXR3b3JrIiwid3JpdGU6bWV0cmljcyIsInJlYWQ6ZXdiIl19.ay_YTwRsfcNzVdmQ4EgmuNMMypfZIIc8K9dCCtLqUmUJDtE7NUuKaVAmGDdmW1J-ngm0UsH4k6B5QpPIJnLIROpdDf7aRzdE9hNFuSHR3arpyCzmO2-TiFDZLFXQjHf0Q-BaxGoXLQBupGYuQaG_3flaLPB3hPV0nqPoBTIoJgG8n2w0Uo2tePe_y2Blqco1sK2wElwyMlYc-UuTyFSvwKlpSXYmO4ppVmbAa9lS2ley6lcv2TwXLCk0KfIIH2E5OBvJHevZqYEzFBAeLCnahKoWxexsVvEfZr40Nhc6oPRT5yJfHRBnCrDnO1fE96rqguQpsDG-HWCtd2GkpnAXNg"
```
[the remark this answers is not in the document]
Comment (Lax125) [reply test/client/test_token_fetcher.py:17]: @sonatype-lift ignore

File: test/client/test_token_fetcher.py [hunk 3]
```diff
@@ -65,41 +69,43 @@ def test_create_token_fetcher_no_auth(mock_get):
...
 
 
-class TestZepbentoken_fetcher:
+class TestZepbenTokenFetcher:
 
-    @mock.patch('zepben.auth.client.token_fetcher.requests.post', side_effect=lambda *args, **kwargs: MockResponse(
+    @mock.patch('zepben.auth.client.zepben_token_fetcher.requests.post', side_effect=lambda *args, **kwargs: MockResponse(
         {"access_token": TOKEN, "refresh_token": mock_refresh_token, "token_type": "Bearer"}, 200))
     def test_fetch_token_successful(self, mock_post):
         token_fetcher = ZepbenTokenFetcher(
```
[the remark this answers is not in the document]
Comment (Lax125) [reply test/client/test_token_fetcher.py:100]: @sonatype-lift ignore

File: src/zepben/auth/common/auth_method.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+__all__ = ["AuthMethod"]
+
+
+class AuthMethod(Enum):
+    """
+    An enum class that represents the different authentication methods that could be returned from the server's
+    ewb/config/auth endpoint.
+    """
+    @classmethod
+    def _missing_(cls, value: str):
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/auth/common/auth_method.py:17]: @sonatype-lift ignore

File: src/zepben/auth/client/zepben_token_fetcher.py [hunk 5]
```diff
@@ -99,96 +83,96 @@ def fetch_token(self) -> str:
...
-            except ValueError as e:
-                raise ValueError(f"Expected JSON response from {url}, but got: {response.text}.", e)
         else:
-            raise ValueError("\n".join(url_err))
+            if response.ok:
+                try:
+                    auth_config_json = response.json()
+                    auth_method = AuthMethod(auth_config_json[auth_type_field])
+                    if auth_method is not AuthMethod.NONE:
+                        return ZepbenTokenFetcher(
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/auth/client/zepben_token_fetcher.py:161]: @sonatype-lift ignore

File: test/client/test_util.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+#  Copyright $year Zeppelin Bend Pty Ltd
+#
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+from zepben.auth import construct_url
+
+assert construct_url("htcpcp", "java", )
```
[the remark this answers is not in the document]
Comment (Lax125) [reply test/client/test_util.py]: @sonatype-lift ignore

File: test/client/test_token_fetcher.py [hunk 7]
```diff
@@ -39,10 +43,10 @@ def json(self):
```
[the remark this answers is not in the document]
Comment (Lax125) [reply test/client/test_token_fetcher.py]: @sonatype-lift ignore

File: test/common/test_status_code.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+from zepben.auth.common.status_code import StatusCode
+
+
+def test_status_code():
+    assert len(StatusCode) == 5
+
+    assert StatusCode.OK.value == 200
+    assert StatusCode.MALFORMED_TOKEN.value == 400
+    assert StatusCode.UNAUTHENTICATED.value == 403
+    assert StatusCode.PERMISSION_DENIED is StatusCode.UNAUTHENTICATED
```
Comment (Lax125) [inline test/common/test_status_code.py]: Just to confirm, do we want these to be aliases? They won't be distinguishable if so.

Comment (Lax125) [reply test/common/test_status_code.py]: (They are both set to 403 in the JVM version too)

File: src/zepben/auth/common/status_code.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#  Copyright $year Zeppelin Bend Pty Ltd
```
Comment (kgreav) [inline src/zepben/auth/common/status_code.py]: $year - fix others as well. just set them to 2022

File: test/client/test_token_fetcher.py [hunk 10]
```diff
@@ -1,4 +1,8 @@
-#  Copyright 2022 Zeppelin Bend Pty Ltd
+#  Copyright $year Zeppelin Bend Pty Ltd
+#
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
```
Comment (kgreav) [inline test/client/test_token_fetcher.py]: duplicate copyright... looks like the IDE is doing the wrong thing with copyrights... damnit.

File: test/common/test_status_code.py (hunk 8 above)
Comment (kgreav) [reply test/common/test_status_code.py]: Is one of them meant to be 401? PERMISSION_DENIED feels like it should be 401?

What's the purpose of using our own StatusCode? Why did you choose that over just using grpc.StatusCode?

**Review by kgreav — CHANGES_REQUESTED** [review]

File: test/common/test_status_code.py (hunk 8 above)
Comment (Lax125) [reply test/common/test_status_code.py]: Honestly I was just blindly transliterating. I'll use grpc.StatusCode instead.
Edit: The `StatusCode` class is really only used in `com.zepben.auth.client` to check if responses are `OK`, and the `requests` library has that covered for us.

Comment (Lax125) [reply test/common/test_status_code.py]: Class removed.

File: src/zepben/auth/common/status_code.py (hunk 9 above)
Comment (Lax125) [reply src/zepben/auth/common/status_code.py]: Used Ctrl+Shift+R to replace `$year` to `2022`.

File: test/client/test_token_fetcher.py (hunk 10 above)
Comment (Lax125) [reply test/client/test_token_fetcher.py]: Fixed and checked other files for occurrences (found 0).

**Pull request merged**